consensus/polybft/extra.go

```diff
@@ -0,0 +1,390 @@
+package polybft
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/go-hclog"
+
+	bls "github.com/0xPolygon/polygon-edge/consensus/polybft/signer"
+
+	"github.com/0xPolygon/polygon-edge/consensus/polybft/bitmap"
+	"github.com/0xPolygon/polygon-edge/types"
+	"github.com/umbracle/ethgo"
+	"github.com/umbracle/fastrlp"
+)
+
+const (
+	// ExtraVanity represents a fixed number of extra-data bytes reserved for proposer vanity
+	ExtraVanity = 32
+
+	// ExtraSeal represents the fixed number of extra-data bytes reserved for proposer seal
+	ExtraSeal = 65
+)
+
+var (
+	// PolyMixDigest represents a hash of "PolyBFT Mix" to identify whether the block is from PolyBFT consensus engine
+	PolyMixDigest = types.Hash(ethgo.HexToHash("adce6e5230abe012342a44e4e9b6d05997d6f015387ae0e59be924afc7ec70c1"))
+)
+
+// Extra defines the structure of the extra field for Istanbul
+type Extra struct {
+	Validators *ValidatorSetDelta
+	Seal       []byte
+	Parent     *Signature
+	Committed  *Signature
+}
+
+// MarshalRLPTo defines the marshal function wrapper for Extra
+func (i *Extra) MarshalRLPTo(dst []byte) []byte {
+	ar := &fastrlp.Arena{}
+
+	return i.MarshalRLPWith(ar).MarshalTo(dst)
+}
+
+// MarshalRLPWith defines the marshal function implementation for Extra
+func (i *Extra) MarshalRLPWith(ar *fastrlp.Arena) *fastrlp.Value {
+	vv := ar.NewArray()
+
+	// Validators
+	if i.Validators == nil {
+		vv.Set(ar.NewNullArray())
+	} else {
+		vv.Set(i.Validators.MarshalRLPWith(ar))
+	}
+
+	// Seal
+	if len(i.Seal) == 0 {
+		vv.Set(ar.NewNull())
+	} else {
+		vv.Set(ar.NewBytes(i.Seal))
+	}
+
+	// ParentSeal
+	if i.Parent == nil {
+		vv.Set(ar.NewNullArray())
+	} else {
+		vv.Set(i.Parent.MarshalRLPWith(ar))
+	}
+
+	// CommittedSeal
+	if i.Committed == nil {
+		vv.Set(ar.NewNullArray())
+	} else {
+		vv.Set(i.Committed.MarshalRLPWith(ar))
+	}
+
+	return vv
+}
+
+// UnmarshalRLP defines the unmarshal function wrapper for Extra
+func (i *Extra) UnmarshalRLP(input []byte) error {
+	return fastrlp.UnmarshalRLP(input, i)
+}
+
+// UnmarshalRLPWith defines the unmarshal implementation for Extra
+func (i *Extra) UnmarshalRLPWith(v *fastrlp.Value) error {
+	elems, err := v.GetElems()
+	if err != nil {
+		return err
+	}
+
+	if num := len(elems); num != 4 {
+		return fmt.Errorf("not enough elements to decode extra, expected 4 but found %d", num)
+	}
+
+	// Validators
+	{
+		if elems[0].Elems() > 0 {
+			i.Validators = &ValidatorSetDelta{}
+			if err := i.Validators.UnmarshalRLPWith(elems[0]); err != nil {
+				return err
+			}
+		}
+	}
```

Comment:
Why do you need to separate this out into a code block like this?
Same question for the remainder of the method

Reply:
It is done like that, because `Extra` has multiple fields which needs to be RLP (de)serialized and therefore it is easier tracking which part of code maps to which field in the `Extra` struct. Agreed that there is no point in doing so from functional side, but it is there just for easier tracking which part of code deals with which part of `Extra` struct.